### ISSUE TITLE
Flawed naming conventions

### ISSUE THREAD

**Issue by thekalinga** [reactor/reactor-netty#561]
Take a look at this API

```java
client
  .headers(builder -> {
    builder.add(COOKIE, new DefaultCookie(SESSION_ID, "hello"));
  })
  .post()
  .uri("/uri")
  .send(ByteBufFlux.fromString(just("hi")))
  .response((httpClientResponse, byteBufFlux) -> {
    return just(sessionIdAndApiKey);
  });
```

When you look at the method `post()`, your intuition tells that the `POST` call is being made. But as per API, you are building a request that uses http `POST` method when the actual `send` method is executed

This is a flawed api design as it breaks intuition. The same is true for all other http methods aswell. This flaw first appeared in `WebClient` API of Webflux & is now carried to reactor-netty aswell

The API needs to make sure that any action methods `get()`, `post()`, e.t.c would be considered by most developers as ACTION methods, not builder methods

A suggestion I have is to name these methods `use<Op>` (or) `op<Builder>()` (or) `for<Op>`, i.e `client.useGet()` (or) `client.getBuilder()` (or) `client.forGet()`

**Comment by thekalinga** [comment]
Had the same issue [Spring Data R2DBC](https://github.com/spring-projects/spring-data-r2dbc/issues/43) aswell

**Comment by thekalinga** [comment]
Also, `send` too feels like an ACTION, but the actual call happens when we subscribe like any other reactive stream. So, even `send` method needs to be renamed to something which implies request building is over & we want to proceed with response handling

**Comment by smaldini** [comment]
We followed the same API guidelines than reactor and transitively functional API especially rx.
Would you think that Flux.map() directly maps something ? or Mono.merge(a, b) ?
The consistency you see between projects is definitively on purpose as we are trying to avoid different API model mindsets where possible to make the whole thing easier to learn and port between projects.

**Comment by thekalinga** [comment]
@smaldini I agree with you that we need some guidelines on naming methods in reactive world so that regular java users do not feel they are caught off guard. Instead of whitelist method of naming things, may be we need rules of what kind of names to avoid (blacklist)

I have few scattered thoughts on this. One such scattered thought is

If the method looks like an imminent action & also it affects how the user perceives about the method's behaviour based on his past imperative model (I agree we are not in imperative model, but most of us are coming to reactive model from this old notions & the method names in imperative model says what it does, here `.get()` does not actually get anything), we need to be cautious about its name

For e.g, take your examples of `flux.map(..)` & `merge(..)`. In these examples, does the methods `map` & `merge` affect how the user's perception about when `map` & `merge` happens? Even if does affect the when, will this have any impact in user's mental model about how this affects the behaviour of the method chain

`x.get()`, `.post()`, `.send()`, `.execute()`, `.fetch()`, `.run()` feels imminent & a sense of immediacy to them, unlike `map` & `merge`. I am sure these are my subjective feelings & others might not feel the same

Plus, once we have fibres in Java 12+, we would want to retain the immediacy of the above terms to mean immediate execution in a fibre that does not block the thread

Overall there seems to be a need for us to frame some rule of thumb about how not to name a method in reactive world as imperative world is not going away anytime soon

**Comment by mp911de** [comment]
> regular java users do not feel 

I think this is the underlying assumption in the whole discussion. Regular translates to imperative programming.


> If the method looks like an imminent action … affect how the user's perception about when (map & merge) happens?

This again depends on the context. There is a point as `.block()` or `.subscribe()` aren't substantially different from a name like `.send()`. The only thing that distinguishes here is _knowledge_ how Project Reactor building blocks are supposed to work.

I also do not think there is really anything that we can do about it without breaking the entire API. Maybe we can take away that `execute()` and `run()` (verbs strongly related to execution/invocation) should be considered more carefully when building new reactive API.

**Comment by thekalinga** [comment]
`.block()` does what it implies from method name. It blocks current thread
`.subscribe()` does it what implies based on method name. It subscribes when the method gets executed

but `.send()` does not actually send. I think this is where the grey area is

> The only thing that distinguishes here is knowledge how Project Reactor building blocks are supposed to work.

I too agree with this statement partially. But only partially as I mentioned above

**Comment by thekalinga** [comment]
A rough guideline I've in my mind is this

"Method names should say what they do, not when they do it"

**Comment by thekalinga** [comment]
Method names like `execute`, `fetch`  (r2dbc), `get`, `post` (the current library), e.t.c breaks this guideline as they implies both `what` and `when`

Avoid method names that imply `when` along with `what`, unless there are no better names

Please note that these are my rough thoughts & I might be completely misguided on these guidelines. I'm only trying to see if we can make life of the rest of 95-99% imperative programmer's life simple

**Comment by thekalinga** [comment]
Also, method names that imply `when` along with `what` that are on streams is somewhat understandable as there is a ground rule which says `subscribe` is the entry point

But, Builders are not streams. So having a method name imply when within builder method is even more 

Applying this `when` vs `what` guideline to adba method names `rowOperation` & `countOperation` indicate that they just imply `what` they do, and does not imply when they do it. Instead of `countOperation` if they named the method `count`, it would have implied `when` aswell. I am not sure if this indeed is the actual reason/there is something else on why adba named `countOperation` instead of just `count`
